### PR TITLE
fix(web): prevent each_key_duplicate when JSONL has duplicate uuids

### LINKED ISSUE #137: Handle duplicate message UUIDs in Svelte MessageList
### Problem

A fatal Svelte runtime error occurs when viewing certain sessions in the Web UI:
`Uncaught Error: https://svelte.dev/e/each_key_duplicate`

This happens at `MessageList.svelte` because the Svelte `{#each}` block uses `msg.uuid` as the uniqueness key:

```svelte
{#each messages as msg, i (msg.uuid ?? `idx-${i}`)}
```

When the underlying `.jsonl` session file contains multiple records with the same `uuid` (which can happen with file synchronization conflicts, repeated history appends, or edits across sessions), Svelte encounters duplicate keys and crashes the page (white screen).

#### Phase 1: UI Hotfix (Immediate Stability)

- [x] Modify `MessageList.svelte` to guarantee unique keys (final implementation: O(n) Set-based first-occurrence keying — preserves stable Svelte identity for the no-duplicate case, only deduplicates true duplicates) — merged in PR #183 (commit `29876ad`):
  ```svelte
  {#each messages as msg, i (msg.uuid ? `${msg.uuid}-${i}` : `idx-${i}`)}
  ```

#### Phase 2: UI Duplicate Warning (Visibility)

- [ ] In `SessionViewer.svelte` or `MessageList.svelte`, proactively scan the `messages` array for duplicate `uuid`s.
- [ ] If duplicates are found, render a non-intrusive warning banner at the top of the session view:
  > Warning: This session file contains duplicate message entries. Please check the raw `.jsonl` file or run session repair.

#### Phase 3: Backend Deduplication (Data Integrity)

- [x] Implement a deduplication filter in `@claude-sessions/core` when parsing JSONL files. — merged in PR #184 (commit `331748b6`)
- [x] `crud.ts:readSession` should deduplicate messages by `uuid`, keeping the latest (last appearing) message for a given `uuid`. — merged in PR #184
- [x] This avoids rendering hundreds of redundant duplicate messages, reducing DOM cost and token usage metrics. — merged in PR #184

### Verification

- Construct a test fixture `.jsonl` with intentionally duplicated lines (same `uuid`).
- Verify the Web UI loads normally without throwing `each_key_duplicate`.
- Verify the backend deduplication correctly filters out the duplicated messages from the array returned by `readSession`.

### PR DESCRIPTION
**Pull request by DrumRobot**
## Summary

Phase 1 UI hotfix for #137. The MessageList `{#each}` block keyed each
record solely on `msg.uuid`, which crashes Svelte with
`each_key_duplicate` whenever the underlying JSONL contains records that
share the same `uuid` (file-sync conflicts, repeated history appends,
cross-session edits). Appending the array index keeps the key unique
while still letting Svelte track items by their stable uuid prefix.

```diff
-    {#each messages as msg, i (msg.uuid ?? `idx-${i}`)}
+    {#each messages as msg, i (msg.uuid ? `${msg.uuid}-${i}` : `idx-${i}`)}
```

A new Playwright regression (`duplicate-uuid-handling.spec.ts`) loads a
fixture whose JSONL intentionally repeats `dup-a` / `dup-b` uuids and
asserts no `each_key_duplicate` error reaches the page console.

Phase 2 (warning banner in `SessionViewer`) and Phase 3 (backend dedup
in `@claude-sessions/core`) are deferred to follow-up PRs.

## Test plan

- [x] `npx playwright test e2e/duplicate-uuid-handling.spec.ts` passes locally (4.4s, 1/1)
- [x] CI build + typecheck pass on this branch (commit 037d844: e2e ✅ 1m10s, test ubuntu ✅ 2m38s, test windows ✅ 4m3s, CodeRabbit ✅)
- [x] Remaining e2e specs unaffected (CI e2e suite green on commit 037d844)
- [x] Open the duplicate-uuid fixture session in a local `pnpm dev` and confirm no white screen (covered by Playwright e2e — runs against the same dev server via CLAUDE_SESSIONS_DIR fixture, asserts no `each_key_duplicate` + Messages tab renders)

Relates to #137

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed message rendering in edge case scenarios to prevent display errors.

* **Tests**
  * Added regression test fixtures and end-to-end tests to verify message handling works correctly in edge case scenarios.

### CODE REVIEW

File: packages/web/e2e/duplicate-uuid-handling.spec.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    await page.goto(`/session/${PROJECT}/${SESSION}`)
+    await page.waitForSelector('button:has-text("Messages")', { timeout: 10000 })
+    await page.waitForLoadState('networkidle')
+
+    const duplicateKeyErrors = errors.filter((m) => m.includes('each_key_duplicate'))
+    expect(
+      duplicateKeyErrors,
+      `Expected no each_key_duplicate errors, got: ${duplicateKeyErrors.join('\n')}`
+    ).toHaveLength(0)
```
Comment (DrumRobot) [inline packages/web/e2e/duplicate-uuid-handling.spec.ts:30]: **🛠️ Refactor | 🟡 Important** — TDD Red phase fails via `waitForSelector` timeout (line 21), never reaching this assertion. A future regression with a different failure mode could be reported as a misleading "button not found" failure rather than a direct `each_key_duplicate` capture.

Suggested strengthening (verified locally to catch the regression directly):

```ts
await page.waitForLoadState('networkidle')

const duplicateKeyErrors = errors.filter((m) => m.includes('each_key_duplicate'))
expect(
  duplicateKeyErrors,
  `Expected no each_key_duplicate errors, got:\n${duplicateKeyErrors.join('\n')}`
).toHaveLength(0)

await expect(page.getByRole('button', { name: /💬 Messages \(\d+\)/ })).toBeVisible({
  timeout: 5000,
})
```

Drop the early `waitForSelector('button:has-text("Messages")')` — that selector also resolves to 2 elements (Navigation mode button + Messages tab), triggering Playwright strict-mode violations. The regex-bound `getByRole` matcher targets only the tab and surfaces the message count for cross-check.

**Review by DrumRobot — COMMENTED** [review]
## Internal Code Review — [receiving-code-review](https://skills.sh/obra/superpowers/receiving-code-review)

### 🛠️ Refactor | 🟡 Important

**Finding 1 — `packages/web/e2e/duplicate-uuid-handling.spec.ts`: regression test does not directly capture `each_key_duplicate`**

TDD Red-Green-Refactor verification was run locally with this commit (`acb993d`) on the duplicate-UUID fixture session:

| Phase | Code | Result | Signal |
|-------|------|--------|--------|
| GREEN (fix applied) | `(msg.uuid ? \`${msg.uuid}-${i}\` : \`idx-${i}\`)` | ✅ PASS (1.5s) | no `each_key_duplicate` + `💬 Messages (7)` tab visible |
| RED (fix reverted) | `(msg.uuid ?? \`idx-${i}\`)` | ❌ FAIL via `waitForSelector` timeout, **not** via the `each_key_duplicate` assertion | page stuck at `Loading…` because hydration crashed before the Messages button could render |
| GREEN (fix restored) | fix applied | ✅ PASS | regression test re-confirmed |

The current spec catches the regression only as a side-effect (`waitForSelector` times out because hydration never finishes). It never reaches the `expect(duplicateKeyErrors).toHaveLength(0)` assertion, so a future regression with a different failure mode could slip through as a misleading "button not found" failure.

**Suggested strengthening** (verified locally to catch the regression directly — RED phase now produces `Expected no each_key_duplicate errors, got: ["each_key_duplicate", "https://svelte.dev/e/each_key_duplicate"]`):

```ts
await page.goto(`/session/${PROJECT}/${SESSION}`)
// Allow hydration + initial fetch. We deliberately avoid waiting on the
// "Messages" button because, without the fix, Svelte crashes mid-hydration
// and the button is never rendered — we want to assert on the *error
// signal*, not the absence of a button.
await page.waitForLoadState('networkidle')

const duplicateKeyErrors = errors.filter((m) => m.includes('each_key_duplicate'))
expect(
  duplicateKeyErrors,
  `Expected no each_key_duplicate errors, got:\n${duplicateKeyErrors.join('\n')}`
).toHaveLength(0)

// Secondary signal: with the fix, the messages render and the Messages
// tab button (with the per-fixture message count) becomes visible.
await expect(page.getByRole('button', { name: /💬 Messages \(\d+\)/ })).toBeVisible({
  timeout: 5000,
})
```

### Verdict

**No — with fixes.** The fix itself (1-line `{#each}` key change) is correct and verified to resolve the crash; only the spec needs hardening so it stays a meaningful regression gate.

### Reviewer

Internal Review fallback (Copilot unavailable for this org per `gh api orgs/es6kr/copilot/billing` → `active: 0, management: "disabled"`).

File: packages/web/src/lib/components/MessageList.svelte [hunk 2]
```diff
@@ -36,7 +36,7 @@
     : 'border border-gh-border rounded-lg'}"
 >
   <div class="{enableScroll ? 'overflow-y-auto' : ''} flex-1 p-4 flex flex-col gap-4">
-    {#each messages as msg, i (msg.uuid ?? `idx-${i}`)}
+    {#each messages as msg, i (msg.uuid ? `${msg.uuid}-${i}` : `idx-${i}`)}
```
Comment (DrumRobot) [inline packages/web/src/lib/components/MessageList.svelte]: **🛠️ Refactor | 🟡 Important** — The `${msg.uuid}-${i}` key guarantees uniqueness (solves the crash) but ties Svelte identity to array position. When a message is deleted mid-list, every subsequent component unmounts + remounts instead of patching in place — visible flash + lost internal state.

**Proposed mitigation** (O(n) Set-based, first-occurrence preserves stable identity):

```svelte
<script lang="ts">
  // …existing props…
  const messageKeys = $derived.by(() => {
    const seen = new Set<string>()
    return messages.map((msg, i) => {
      if (!msg.uuid) return `idx-${i}`
      if (!seen.has(msg.uuid)) { seen.add(msg.uuid); return msg.uuid }
      return `${msg.uuid}-${i}`
    })
  })
</script>

{#each messages as msg, i (messageKeys[i])}
```

Acceptable hotfix without this mitigation IF the trade-off is documented in the PR description + a Phase 3 follow-up tracks the stable-identity refinement.

**Review by DrumRobot — COMMENTED** [review]
## Internal Code Review — [requesting-code-review](https://skills.sh/obra/superpowers/requesting-code-review)

> Second-pass review via the `code-reviewer` subagent. Adds one finding the prior self-review missed (stable Svelte identity trade-off) and corroborates the spec-strengthening finding already posted in review [4493077374](https://github.com/es6kr/claude-code-sessions/pull/183#pullrequestreview-4493077374).

### Strengths

- Minimal scope — Phase 1 only, Phase 2/3 cleanly deferred
- Mathematical correctness — new key is unique unconditionally (`A-0` ≠ `A-1`)
- Common-case behavior unchanged for non-duplicate sessions
- Fixture realism — models sync conflict + normal continuation per Issue #137
- Fixture naming follows `-Users-test-<name>` convention
- Existing e2e specs (`project-tree-grouped`, `rename-session`, `session-title`, `split-session-message-sync`) unaffected
- Hydration wait present (`waitForLoadState('networkidle')`)

### Findings

#### 🛠️ Refactor | 🟡 Important — `MessageList.svelte:39` stable Svelte identity erased for the no-duplicate case

When messages have unique uuids (the normal case), the old key was just `msg.uuid` — a stable identifier that travels with the message regardless of array position. The new key is `${msg.uuid}-${i}`, which ties identity to position. If the parent deletes a message at index 3, every message from index 4 onward gets a new key, and Svelte unmounts + remounts those components instead of patching them in place. In a long session this causes a visible flash and discards any internal component state.

Not a crash — but it degrades the reconciliation quality the keyed `{#each}` was designed to provide. Trade-off: the fix solves the crash unconditionally, at the cost of degraded patch efficiency when uuids are unique but a message is removed mid-list.

**Proposed mitigation** (O(n) Set-based first-occurrence keying — preserves stable identity when uuids are unique, only deduplicates true duplicates):

```svelte
<script lang="ts">
  // …existing props…

  const messageKeys = $derived.by(() => {
    const seen = new Set<string>()
    return messages.map((msg, i) => {
      if (!msg.uuid) return `idx-${i}`
      if (!seen.has(msg.uuid)) {
        seen.add(msg.uuid)
        return msg.uuid          // first occurrence: stable identity
      }
      return `${msg.uuid}-${i}`  // subsequent occurrence: deduplicate
    })
  })
</script>

<!-- …existing markup… -->
{#each messages as msg, i (messageKeys[i])}
```

**Acceptable hotfix without this mitigation IF** the trade-off is documented in the PR description and a Phase 3 follow-up tracks the stable-identity refinement. Phase 3 backend dedup will eliminate true duplicates upstream, after which the simple `msg.uuid` key would suffice again — but defense-in-depth (keep the index suffix as a fallback) is still the safer posture.

#### 🛠️ Refactor | 📝 Minor — `duplicate-uuid-handling.spec.ts:21` RED phase fails via `waitForSelector` timeout, not via the direct assertion

Confirms the self-review finding in review [4493077374](https://github.com/es6kr/claude-code-sessions/pull/183#pullrequestreview-4493077374). Replace `waitForSelector('button:has-text("Messages")')` with `waitForLoadState('networkidle')` so the spec settles regardless of hydration outcome, then assert on `each_key_duplicate` keyword presence; promote the Messages-button visibility to a secondary check using the more specific `getByRole({ name: /💬 Messages \(\d+\)/ })` matcher to avoid strict-mode violation against the Navigation-mode button.

Verified locally (RED phase now produces `Expected no each_key_duplicate errors, got: ["each_key_duplicate", "https://svelte.dev/e/each_key_duplicate"]` — direct assertion failure, not a timeout).

#### 📝 Nitpick | 📝 Minor — `session-dup-uuid.jsonl:1` summary record carries a `timestamp` field

`CLAUDE.md` states: *"Summary records have `leafUuid` but no timestamp."* The new fixture's line 1 has both `leafUuid` and `timestamp`. **Pre-existing inconsistency** — the older `session-with-summary.jsonl` fixture also includes a timestamp on its summary record. No functional impact on this PR (the spec never reads summary content); flagging for awareness only.

### Architecture / Phase Split

The Phase 1/2/3 split is correct and should not be collapsed. Phase 2 (warning banner) and Phase 3 (backend dedup) each introduce non-trivial UI work and core library changes; merging them into the hotfix would increase risk surface. Defense-in-depth: even after Phase 3 lands and `readSession` dedups, keeping the index suffix in the MessageList key removes silent dependency on Phase 3 correctness.

### Backward Compatibility

None. Svelte keys exist only at runtime during reconciliation — no serialized state. Changing the key format is safe.

### Assessment

**Ready to merge: With fixes**

One-line fix is correct and sufficient for Phase 1. Recommend strengthening the spec (Minor [2]) in this same PR — its sole purpose is to be a regression gate. The stable-identity trade-off (Important [1]) is acceptable for a hotfix when documented; the O(n) Set-based key is the cleaner long-term shape.

### Reviewer matrix

> **CodeRabbit** (walkthrough — `No actionable comments`) + **Internal Code Review** (Copilot unavailable on `es6kr` org — `gh api orgs/es6kr/copilot/billing` returns `active: 0, management: disabled` — auto-fallback per `consolidate/pr.md` Step 2.4)
